components/server/src/api/server.ts

```diff
@@ -123,55 +138,78 @@ export class API {
                         grpc_type = "bidi_stream";
                     }
 
+                    logContext.requestId = v4();
+
                     grpcServerStarted.labels(grpc_service, grpc_method, grpc_type).inc();
                     const stopTimer = grpcServerHandling.startTimer({ grpc_service, grpc_method, grpc_type });
                     const done = (err?: ConnectError) => {
                         const grpc_code = err ? Code[err.code] : "OK";
                         grpcServerHandled.labels(grpc_service, grpc_method, grpc_type, grpc_code).inc();
                         stopTimer({ grpc_code });
+                        let callMs: number | undefined;
+                        if (callStartedAt) {
+                            callMs = performance.now() - callStartedAt;
+                        }
+                        log.debug("public api: done", { grpc_code, verifyMs, callMs });
+                        // right now p99 for getLoggetInUser is around 100ms, using it as a threshold for now
+                        const slowThreshold = 100;
+                        const totalMs = performance.now() - logContext.contextTimeMs;
+                        if (grpc_type === "unary" && totalMs > slowThreshold) {
+                            log.warn("public api: slow unary call", { grpc_code, callMs, verifyMs });
+                        }
                     };
                     const handleError = (reason: unknown) => {
                         let err = ConnectError.from(reason, Code.Internal);
                         if (reason != err && err.code === Code.Internal) {
-                            console.error("public api: unexpected internal error", reason);
-                            // don't leak internal errors to a user
-                            // TODO(ak) instead surface request id
-                            err = ConnectError.from(`please check server logs`, Code.Internal);
+                            log.error("public api: unexpected internal error", reason);
+                            err = ConnectError.from(
+                                `Oops! Something went wrong. Please quote the request ID ${logContext.requestId} when reaching out to Gitpod Support.`,
+                                Code.Internal,
+                            );
                         }
                         done(err);
                         throw err;
                     };
 
+                    let verifyMs: number | undefined;
+                    let callStartedAt: number | undefined;
                     const context = args[1] as HandlerContext;
-                    async function call<T>(): Promise<T> {
+
+                    const apply = async <T>(): Promise<T> => {
+                        const verifyStartedAt = performance.now();
                         const user = await self.verify(context);
```

Comment:
I was wondering rather add a warning and more fine grain logging inside verify generally. But i worry about negative impact form logging itself. It is probably the hottest path.